packages/kbn-test/src/functional_test_runner/lib/suite_tracker.ts

```diff
@@ -0,0 +1,146 @@
+  leafSuite: boolean;
+}
+
+const getTestMetadataPath = () => {
+  return process.env.TEST_METADATA_PATH || resolve(REPO_ROOT, 'target', 'test_metadata.json');
+};
+
+export class SuiteTracker {
+  finishedSuitesByConfig: Record<string, Record<string, SuiteWithMetadata>> = {};
+  inProgressSuites: Map<object, SuiteInProgress> = new Map<object, SuiteInProgress>();
+
+  static startTracking(lifecycle: Lifecycle, configPath: string): SuiteTracker {
+    const suiteTracker = new SuiteTracker(lifecycle, configPath);
+    return suiteTracker;
+  }
+
+  getTracked(suite: object): SuiteInProgress {
+    if (!this.inProgressSuites.has(suite)) {
+      this.inProgressSuites.set(suite, { success: undefined } as SuiteInProgress);
+    }
+    return this.inProgressSuites.get(suite)!;
+  }
+
+  constructor(lifecycle: Lifecycle, configPathAbsolute: string) {
+    if (fs.existsSync(getTestMetadataPath())) {
+      fs.unlinkSync(getTestMetadataPath());
+    } else {
+      fs.mkdirSync(dirname(getTestMetadataPath()), { recursive: true });
+    }
+
+    const config = relative(REPO_ROOT, configPathAbsolute);
+
+    lifecycle.beforeTestSuite.add(suite => {
+      const tracked = this.getTracked(suite);
+      tracked.startTime = new Date();
+    });
+
+    // If a test fails, we want to make sure all of the ancestors, all the way up to the root, get marked as failed
+    // This information is not available on the mocha objects without traversing all descendants of a given node
+    const handleFailure = (_: any, test: any) => {
+      let parent = test.parent;
+
+      // Infinite loop protection, just in case
+      for (let i = 0; i < 500 && parent; i++) {
+        if (this.inProgressSuites.has(parent)) {
+          this.getTracked(parent).success = false;
+        }
+        parent = parent.parent;
+      }
+    };
+
+    lifecycle.testFailure.add(handleFailure);
+    lifecycle.testHookFailure.add(handleFailure);
+
+    lifecycle.afterTestSuite.add(suite => {
+      const tracked = this.getTracked(suite);
+      tracked.endTime = new Date();
+
+      // The suite ended without any children failing, so we can mark it as successful
+      if (typeof tracked.success === 'undefined') {
+        tracked.success = true;
+      }
+
+      let duration = tracked.endTime.getTime() - (tracked.startTime || new Date()).getTime();
+      duration = Math.floor(duration / 1000);
+
+      const file = relative(REPO_ROOT, suite.file);
+
+      this.finishedSuitesByConfig[config] = this.finishedSuitesByConfig[config] || {};
+
+      // This will get called multiple times for a test file that has multiple describes in it or similar
+      // This is okay, because the last one that fires is always the root of the file, which is is the one we ultimately want
+      this.finishedSuitesByConfig[config][file] = {
+        ...tracked,
+        duration,
+        config,
+        file,
+        tag: suite.suiteTag,
+        title: suite.title,
+        leafSuite: !!(
```

Comment:
Let's use a better name for this that communicates the fact that this is a flag based on wether the file has any tests. I suggest either `hasTests` or keeping a count of tests and using `testCount: xyz`